src/audio/pipeline/pipeline-graph.c

```diff
@@ -423,29 +423,35 @@ int pipeline_for_each_comp(struct comp_dev *current,
 }
 
 /* visit connected pipeline to find the dai comp */
-struct comp_dev *pipeline_get_dai_comp(uint32_t pipeline_id)
+struct comp_dev *pipeline_get_dai_comp(uint32_t pipeline_id, int dir)
 {
-	struct ipc_comp_dev *sink;
+	struct ipc_comp_dev *crt;
 	struct ipc *ipc = ipc_get();
 
-	sink = ipc_get_ppl_sink_comp(ipc, pipeline_id);
-	while (sink) {
+	crt = ipc_get_ppl_comp(ipc, pipeline_id, dir);
+	while (crt) {
 		struct comp_buffer *buffer;
 		struct comp_dev *comp;
 
-		/* dai is found */
-		if (list_is_empty(&sink->cd->bsink_list))
-			return sink->cd;
+		/* based on given direction check if list of
+		 * sink or source buffers is empty. If so
+		 * then we have found the DAI.
+		 */
+		if ((dir == PPL_DIR_DOWNSTREAM &&
+		     list_is_empty(&crt->cd->bsink_list)) ||
+		    (dir == PPL_DIR_UPSTREAM &&
+		     list_is_empty(&crt->cd->bsource_list)))
+			return crt->cd;
 
-		buffer = buffer_from_list(comp_buffer_list(sink->cd, PPL_DIR_DOWNSTREAM)->next,
-					  struct comp_buffer, PPL_DIR_DOWNSTREAM);
-		comp = buffer_get_comp(buffer, PPL_DIR_DOWNSTREAM);
+		buffer = buffer_from_list(comp_buffer_list(crt->cd, dir)->next,
+					  struct comp_buffer, dir);
```

Comment:
so for forking pipelines, e.g. connected over mux / demux we just get whichever happens to be the first in the buffer list, right?

Reply:
Yeah, that's right. This would mean having 2 or more DAIs in the same pipeline which I think getting the timestamp for the first in the list should be fine. I don't think we have such a pipeline though.